lib/r10k/source/branchname.rb

```diff
@@ -0,0 +1,112 @@
+require 'r10k/git'
+require 'r10k/environment'
+require 'r10k/util/purgeable'
+require 'r10k/util/core_ext/hash_ext'
+
+# This class controls the behaviour of branch names
+#
+# A branch name becomes the name of the environment on the file system.
+# The default branch name is the name of the source branch.
+# The name can be modified with a prefix, and by default will be automatically
+# modified if the branch name contains characters invalid for a file system.
+#
+class R10K::Source::BranchName
```

Comment:
The `R10K::Source` namespace deals with collecting and creating environments, while BranchName is specific to a single environment. I would rather this live in the `R10K::Environment` namespace.